modules/diskq/logqueue-disk-non-reliable.c

```diff
@@ -275,38 +315,113 @@ _pop_head (LogQueueDisk *s, LogPathOptions *path_options)
         {
           msg = g_queue_pop_head (self->qoverflow);
           POINTER_TO_LOG_PATH_OPTIONS (g_queue_pop_head (self->qoverflow), path_options);
-          log_queue_memory_usage_sub(&self->super.super, log_msg_get_size(msg));
+          log_queue_memory_usage_sub(s, log_msg_get_size(msg));
         }
     }
 
   if (msg != NULL)
     {
-      if (self->super.super.use_backlog)
+      if (s->use_backlog)
         {
           log_msg_ref (msg);
           g_queue_push_tail (self->qbacklog, msg);
           g_queue_push_tail (self->qbacklog, LOG_PATH_OPTIONS_TO_POINTER (path_options));
         }
       _move_disk (self);
+      log_queue_queued_messages_dec(s);
     }
+
+  g_static_mutex_unlock(&s->lock);
+
   return msg;
 }
 
 static void
-_push_head (LogQueueDisk *s, LogMessage *msg, const LogPathOptions *path_options)
+_push_head(LogQueue *s, LogMessage *msg, const LogPathOptions *path_options)
 {
-  LogQueueDiskNonReliable *self = (LogQueueDiskNonReliable *) s;
+  LogQueueDiskNonReliable *self = (LogQueueDiskNonReliable *)s;
+
+  g_static_mutex_lock(&s->lock);
+
+  g_queue_push_head(self->qout, LOG_PATH_OPTIONS_TO_POINTER (path_options));
+  g_queue_push_head(self->qout, msg);
+  log_queue_queued_messages_inc(s);
+  log_queue_memory_usage_add(s, log_msg_get_size(msg));
 
-  g_queue_push_head (self->qout, LOG_PATH_OPTIONS_TO_POINTER (path_options));
-  g_queue_push_head (self->qout, msg);
-  log_queue_queued_messages_inc(&self->super.super);
-  log_queue_memory_usage_add(&self->super.super, log_msg_get_size(msg));
+  g_static_mutex_unlock(&s->lock);
+}
+
+static void
+_drop_msg(LogQueueDiskNonReliable *self, LogMessage *msg, const LogPathOptions *path_options)
+{
+  stats_counter_inc(self->super.super.dropped_messages);
+
+  if (path_options->flow_control_requested)
+    log_msg_ack(msg, path_options, AT_SUSPENDED);
+  else
+    log_msg_drop(msg, path_options, AT_PROCESSED);
+}
+
+static inline gboolean
+_need_serialized(LogQueueDiskNonReliable *self)
```

Comment:
```suggestion
/* _is_msg_serialization_needed_hint() must be called without holding the queue's lock.
 * This can only be used _as a hint_ for performance considerations, because as soon as the lock
 * is released, there will be no guarantee that the result of this function remain correct. */
static inline gboolean
_is_msg_serialization_needed_hint(LogQueueDiskNonReliable *self)
```

Reply:
Thanks, added!